### PR TITLE
[DS-317] Allow negative field specifiers

### PR DESCRIPTION
**Pull request by lbam**
# This Pull request contains changes to:

`_field=-foo,-bar` now causes all fields to be returned except for foo and bar.

Combining positive and negative mentions in the fields list is not allowed.

# In case changes new features added

- [x] Customer friendly documentation added into `docs/`.
- [ ] Developer friendly documentation added into `DEVELOPMENT|README.md`

# In case breaking changes introduced into API

- [ ] There is customer notification plan: ....

# In case this PR reverts changes in repo

- [ ] Extra details describing reasoning: ...

### CODE REVIEW

File: src/tests/test_rest_framework_dso/test_serializers.py [hunk 1]
```diff
@@ -137,7 +137,18 @@ def test_pagination_many(drf_request, movie):
...
+    _test_fields_limit(api_rf, movie, "name")
+
+
+@pytest.mark.django_db
+def test_fields_limit_negative(api_rf, movie):
+    """Prove that the serializer can omit output fields on request."""
+    # Omit everything but the name.
+    _test_fields_limit(api_rf, movie, "-category_id")
+
+
```
Comment (vdboor) [inline src/tests/test_rest_framework_dso/test_serializers.py:149]: Use `@pytest.mark.parametrize("fields", ["name", "-category_id"])` for this.

File: src/rest_framework_dso/serializers.py [hunk 2]
```diff
@@ -237,13 +237,30 @@ def get_fields(self):
...
+        # Split into fields to include and fields to omit (-fieldname).
+        display_fields, omit_fields = set(), set()
+        for field in request_fields.split(","):
+            if field.startswith("-"):
+                omit_fields.add(field[1:])
+            else:
+                display_fields.add(field)
+
+        if display_fields and omit_fields:
+            raise ValidationError("fields list must be only positive or only negative")
```
Comment (vdboor) [inline src/rest_framework_dso/serializers.py:250]:
```suggestion
            raise ValidationError("It's not possible to combine inclusions and exclusions in the _fields parameter.")
```

File: src/dso_api/dynamic_api/views/api.py [hunk 3]
```diff
@@ -157,7 +157,10 @@ def _get_viewset_api_docs(
         lines.extend(f"* {name}" for name in embedded_fields)
         lines.append("\nExpand everything using `_expand=true`.")
 
-    lines.append("\nUse `?_fields=field,field2` to limit which fields to receive")
+    lines.append(
+        "\nUse `?_fields=field1,field2` to limit which fields to receive,"
+        "\nor `?_fields=-field3,-field4` to get the default list with some exceptions"
```
Comment (vdboor) [inline src/dso_api/dynamic_api/views/api.py:162]:
```suggestion
        "\nor `?_fields=-field3,-field4` to exclude specific fields from the response."
```

File: src/rest_framework_dso/filters.py [hunk 4]
```diff
@@ -571,10 +571,12 @@ def to_html(self, request, queryset, view):
         """
         return ""
 
+    _NON_FIELDS = {"_fields", "_format", "_sort", "_pageSize", "_page_size"}
+
     def is_unfiltered(self, request):
         # If there are request parameters (except for this hard-coded exclude list),
         # they are assumed to be filters.
-        return {"_format", "_sort", "_pageSize", "_page_size"}.issuperset(request.GET.keys())
+        return request.GET.keys() <= self._NON_FIELDS
```
Comment (vdboor) [inline src/rest_framework_dso/filters.py:579]: ok, cool. hoe werkt dit?

File: src/rest_framework_dso/serializers.py (hunk 2 above)
Comment (lbam) [reply src/rest_framework_dso/serializers.py:250]: Done

File: src/tests/test_rest_framework_dso/test_serializers.py (hunk 1 above)
Comment (lbam) [reply src/tests/test_rest_framework_dso/test_serializers.py]: Done.

File: src/dso_api/dynamic_api/views/api.py (hunk 3 above)
Comment (lbam) [reply src/dso_api/dynamic_api/views/api.py]: done

File: src/rest_framework_dso/filters.py (hunk 4 above)
Comment (lbam) [reply src/rest_framework_dso/filters.py:579]: `set.__le__` is hetzelfde als `set.issubset`:

```python
>>> {1,2,3} <= {1,2,3}
True
>>> {1,2,3} <= {1,2,3,4}
True
>>> {1,2,3} <= {1,2}
False
```

Het scheelt één paar haakjes, en omdat de verzameling groter werd, moest ik de regel toch al opbreken.

File: src/rest_framework_dso/serializers.py [hunk 5]
```diff
@@ -237,13 +237,32 @@ def get_fields(self):
...
+            invalid_fields = omit_fields - fields
+        else:
+            display_fields = display_fields
+            invalid_fields = display_fields - fields
+
         if invalid_fields:
             # Some of `display_fields` are not in result.
             raise ValidationError(
-                "'{}' is not one of available options".format("', '".join(sorted(invalid_fields))),
+                "'{}' not among the available options".format("', '".join(sorted(invalid_fields))),
```
Comment (gkoller) [inline src/rest_framework_dso/serializers.py:265]: Is this good?
```python
"', '".join(sorted(invalid_fields))
```
I would have expected:
```python
", ".join(sorted(invalid_fields))
```

Comment (lbam) [reply src/rest_framework_dso/serializers.py:265]: Note the '' around {}:

```python
>>> print("'{}' bla bla".format("', '".join(["foo", "bar"])))
'foo', 'bar' bla bla
```

File: docs/source/generic/rest.rst [hunk 6]
```diff
@@ -119,7 +119,15 @@ Gebruik de :samp:`?_fields={veld1},{veld2},{...}` parameter om alleen specifieke
 
 .. code-block:: bash
 
-    curl 'https://api.data.amsterdam.nl/v1/bag/stadsdeel/?_fields=id,code,naam'
+    curl 'https://api.data.amsterdam.nl/v1/fietspaaltjes/fietspaaltjes/?fields=geometry,soortPaaltje'
+
+Als de veldnamen voorafgegaan worden door een minteken, dan worden alle velden behalve de genoemde
+opgestuurd:
+
+.. code-block:: bash
```
Comment (gkoller) [inline docs/source/generic/rest.rst:127]: Not sure if `bash` is correct. Sphinx delegates to Pygments and Pygments says: https://pygments.org/languages/#interactive-terminal-shell-sessions

Hence i presume it should either be:
```ReST
.. code-block:: console
```
or
```ReST
.. code-block:: shell-session
```

Comment (gkoller) [reply docs/source/generic/rest.rst:127]: Ah, never mind, `bash` is supported (as is `zsh` and `tcsh`)

**Review by gkoller — APPROVED** [review]

File: docs/source/generic/rest.rst (hunk 6 above)
Comment (lbam) [reply docs/source/generic/rest.rst:127]: I double-checked and it looks fine in the generated HTML, with colored quoted strings and everything :)

**Review by ngaranko — APPROVED** [review]

File: src/rest_framework_dso/serializers.py [hunk 7]
```diff
@@ -237,13 +237,32 @@ def get_fields(self):
...
+            raise ValidationError(
+                "It's not possible to combine inclusions and exclusions in the _fields parameter"
+            )
+
+        fields = set(fields.keys())
+        if omit_fields:
+            display_fields = fields - omit_fields
+            invalid_fields = omit_fields - fields
+        else:
+            display_fields = display_fields
```
Comment (lbam) [inline src/rest_framework_dso/serializers.py:259]: Whoops, removing this one before the merge

**Pull request merged**